gene_expression_processing.py

```diff
@@ -0,0 +1,263 @@
+    celltypes: str,
+    chromosomes: str,
+    gene_info_tsv: str,
+    expression_files_prefix: str,
+    sample_mapping_file_path: str,
+    min_pct_expr: int,
+    cis_window_size: int,
+    max_gene_concurrency=int
+):
+    """
+    Run expression processing pipeline
+    """
+    config = get_config()
+    b = get_batch()
+    
+
+    logging.info(f'Cell types to run: {celltypes}')
+    logging.info(f'Chromosomes to run: {chromosomes}')
+
+    # create phenotype covariate files
+    smf_df = pd.read_csv(sample_mapping_file_path, sep='\t')
+    gene_info_df = pd.read_csv(gene_info_tsv, sep='\t')
+
+    for celltype in celltypes.split(','):
+        # get covariates (cell type specific)
+        cov_df = get_celltype_covariates(
+            expression_files_prefix=expression_files_prefix, cell_type=celltype
+        )
+        for chromosome in chromosomes.split(','):
+            # get expression (cell type + chromosome)
+            #expr_adata = get_chrom_celltype_expression(
+            #    gene_info_df=gene_info_df,
+            #    expression_files_prefix=expression_files_prefix,
+            #    chromosome=chromosome,
+            #    cell_type=celltype
+            #)
+            # remove lowly expressed genes
+            #filter_adata = filter_lowly_expressed_genes(
+            #    expression_adata=expr_adata, min_pct=min_pct_expr
+            #)
+
+            #j = b.new_python_job(name='Get expression (cell type + chr)')
+            #j.storage('20G')
+            #.cpu(8)
+            #j.image(config['workflow']['driver_image'])
+            #filter_adata = j.call(get_chrom_celltype_expression,gene_info_df,expression_files_prefix,chromosome,celltype)
+            
+            #f = b.new_python_job(name = 'filter lowly expressed genes')
+            #f.storage('20G')
+            #f.cpu(8)
+            #f.image(config['workflow']['driver_image'])
+           # filter_adata = f.call(filter_lowly_expressed_genes,expr_adata, min_pct_expr)
+            #b.write_output(filter_adata.as_str(), output_path('hope-test-filter_adata.txt'))
+                
+            expression_h5ad_path = to_path(
+            dataset_path(
+                f'tob_wgs_genetics/saige_qtl/hope-test-input/filtered22.h5ad'
+            )
+             ).copy('here.h5ad')
+            filter_adata = scanpy.read(expression_h5ad_path)
+        
+            # combine files for each gene
+            # pylint: disable=no-member
+            for gene in filter_adata.var_names:
+                pheno_cov_job = b.new_python_job(name=f'Create pheno cov job for {gene}')
+                pheno_cov_job.storage('35G')
+                pheno_cov_job.cpu(8)
+                pheno_cov_job.image(config['workflow']['driver_image'])
+
+                # pass the output file path to the job, don't expect an object back
+                pheno_cov_job.call(
+                    build_pheno_cov_filename,gene,cov_df,filter_adata,smf_df,output_path(
```

Comment:
Rather than passing the objects directly here `cov_df` `filter_adata`, `smf_df`, pass the paths of the files to the function, and make the function read them. That will dramatically reduce the need for Hail batch to pickle objects, and it'll likely be quicker, because reading the file is quicker than unpickling (usually).